### ISSUE TITLE
Building Natron on Linux (CentOS7)

### ISSUE THREAD

**Issue by fonkle** [NatronGitHub/Natron#351]
Building Natron on Linux (CentOS 6.10) from the Natron SDK seems to work using these commands:
```
QT_SELECT=4 qmake -r CONFIG+=enforce-gcc8
make -j6
sudo make install
```
However, if I run either _../App/Natron_ or _/usr/local/bin/Natron_ afterwards, it exits with this error:
```
import NatronEnginePython error:
Python exception: could not import module 'PySide.QtCore'
```
The pyside build in the SDK seems successful.

I seem to be missing some steps? Any help would be appreciated!

**Comment by devernay** [comment]
you must set a few env variables for natron to find the python modules and the openfx plugins (which should be built before, see the docs).
```
PYTHON_HOME="$SDK_HOME"
PYTHON_PATH="$SDK_HOME/lib/python${PYVER}"
```
You may also have to set LD_LIBRARY_PATH to contain /opt/Natron-sdk/lib, /opt/Natron-sdk/ffmpeg-gpl2/lib  /opt/Natron-sdk/libraw-gpl2/lib and maybe a few others.
Do not forget to also pull OpenColorIO-Configs.
You also have to set the two following env variables before running natron:
FONTCONFIG_FILE="$natron/Contents/Resources/etc/fonts/fonts.conf"
OCIO="$natron/Contents/Resources/OpenColorIO-Configs/blender/config.ocio"

Of course, when doing a release, we fix all binaries and add a script so that it is no longer necessary.

Once you've figured out the variables to set, please update the docs accordingly.

**Comment by fonkle** [comment]
I cannot get it to work :(.

I have built the plugins and put OpenColorIO-Configs in place (v2.3), but setting all these env variables make no difference. I have added all possible pyside folders (like /opt/Natron-sdk/lib/qt4/python2.7) to PYTHON PATH and LD_LIBRARY_PATH, but to no avail.

**Comment by fonkle** [comment]
The jenkins script _tools/jenkins/launchBuildMain.sh_ gives an error as well:
```
In file included from ../gmic_plugin.cpp:133:
../gmic_stdlib_gmic.h:2:1: error: expected unqualified-id before '<' token
 <?xml version="1.0" encoding="utf-8"?>
 ^
../gmic_plugin.cpp: In constructor 'PluginGlobalData::PluginGlobalData()':
../gmic_plugin.cpp:170:45: error: 'get_gmic_rc_path' was not declared in this scope
   string effectContent = loadStringFromFile(get_gmic_rc_path() + "gmic_ofx.gmic");
                                             ^~~~~~~~~~~~~~~~
../gmic_plugin.cpp:184:40: error: 'gmic_stdlib_gmic' was not declared in this scope
    effectContent = string((const char*)gmic_stdlib_gmic, gmic_stdlib_gmic_len);
                                        ^~~~~~~~~~~~~~~~
../gmic_plugin.cpp:184:58: error: 'gmic_stdlib_gmic_len' was not declared in this scope
    effectContent = string((const char*)gmic_stdlib_gmic, gmic_stdlib_gmic_len);

```

**Comment by fonkle** [comment]
And this one:
```
In file included from ../../../libs/yaml-cpp/include/yaml-cpp/node/iterator.h:13,
                 from ../../../libs/yaml-cpp/include/yaml-cpp/node/impl.h:11,
                 from ../../../libs/yaml-cpp/include/yaml-cpp/node/convert.h:17,
                 from ../../../libs/yaml-cpp/src/convert.cpp:3:
../../../libs/yaml-cpp/include/yaml-cpp/node/detail/iterator.h: In member function 'void YAML_NATRON::detail::iterator_base<V>::increment()':
../../../libs/yaml-cpp/include/yaml-cpp/node/detail/iterator.h:48:54: error: 'next' is not a member of 'boost'
   void increment() { this->base_reference() = boost::next(this->base()); }

```

**Comment by devernay** [comment]
the issue with gmic_stdlib_gmic.h seems to be a wrong download. check how this file is downloaded


at what stage does the yaml-cpp error occur? It seems to be an old version which still depends on boost

**Comment by fonkle** [comment]
It uses the master branch of NatronGitHub/openfx-gmic, which seems broken (clang) since the update from 2.3.1 to 2.3.3 in july 2018.

The yaml-cpp error occurs in the launchBuildMain.sh script (after restarting from the previous error).

**Comment by devernay** [comment]
openfx-gmic was not broken (take a closer look at the travis builds), but I upgraded it to gmic 2.6.2 anyway. There was probably a network issue when the build tried to download `gmic_stdlib_gmic.h`.
 the yaml-cpp error occurs probably when building something that incudes its own version of yaml-cpp, can you chack when it occurs in the logs? I would guess it's OpenColorIO or something like that.
You can safely upgrade OCIO to its latest release, which fixes many issues. Anyway, we will have to upgrade most packages before starting to test. OpenImageIO must be upgraded too.

**Comment by fonkle** [comment]
Cool, thx! _gmic_ seems to work now; _build-plugins.sh_ finishes without errors.

The _yaml-cpp_ error occurs somewhere in _build-natron.sh_. The _launchBuildMain_ script is not logging to a file, is it?

By upgrading OCIO and OIIO, do you mean in the Natron SDK?

**Comment by fonkle** [comment]
In libs/yaml-cpp version 0.5.3 is included:

> Version 0.5.3 obtained from https://github.com/jbeder/yaml-cpp/releases (only src and include directories)

along with 2 patches:

005a6a19ee0af3e284637abe9ecea5501993fc4c.diff
8c35a8ffabea1c989d9f449bff053081fdc987e1.diff

However Natron SDK seems to prefer 0.6.2 from GCC version 4.x and up:
```
# Install yaml-cpp (0.5.3 requires boost, 0.6+ requires C++11, used by OpenColorIO)
YAMLCPP_VERSION=0.5.3
if [[ ! "$GCC_VERSION" =~ ^4\. ]]; then
    YAMLCPP_VERSION=0.6.2 # 0.6.0 is the first version to require C++11
fi
```
Should it be updated to 0.6.2 in _Natron/libs/yaml-cpp_ as well? That would probably break with GCC < 4.0.

**Comment by fonkle** [comment]
I have replaced _libs/yaml-cpp/include_ and _libs/yaml-cpp/src_ with the ones from version 0.6.2 which I downloaded from https://github.com/jbeder/yaml-cpp/releases. This seems to work (at least build).

Next build error after that is:
```
{standard input}: Assembler messages:
{standard input}:870628: Warning: end of file not at end of a line; newline inserted
g++: fatal error: Killed signal terminated program cc1plus
```

**Comment by devernay** [comment]
while building which file? This looks either like a compiler error or a lack of RAM

**Comment by fonkle** [comment]
It was while compiling eigen3. But since this was before using Docker, let's wait if it surfaces again if built in a Docker container?

**Comment by devernay** [comment]
I've now put a script that helps setting up the current shell for building/running/testing Natron and plugins, using the SDK. It's in [tools/utils/natron-sdk-setup-linux.sh](https://github.com/NatronGitHub/Natron/blob/RB-2.3/tools/utils/natron-sdk-setup-linux.sh)

**Comment by Ivanohe73** [comment]
**Natron-sdk Centos 7.6 build issue and fix** 

I have build the sdk on different centos 7 machines and here are some additional step that are needed : 

* the default behavior of the build script is to clean /tmp but unless you are root, this will silently fail and build won't start. The fix is to do this : 

```
mkdir -p ~/natron_workspace/src
mkdir ~/natron_workspace/tmp

WORKSPACE=~/natron_workspace/ TMPDIR=~/natron_workspace/tmp/ include/scripts/build-Linux-sdk.sh
```

* Perl issue : At some point the build will stop at the “Building qt4webkit…” step. This is because the centos 7 perl version is 5.16 while the Natron builded version is 5.30 or above. But qt4webkit refer to /usr/bin/perl in some place. The default centos 7 perl has no file named version.pm. I tried without succes to update perl on the centos without any luck (it would keep the original /usr/bin/perl and add new perl-version file). Anyway, the work around I found is this : 

```
cd /usr/bin/
sudo mv perl perl_ori
sudo ln -s /opt/Natron-sdk/bin/perl
```

relaunch the build. Once the build is finished, put the original perl.

```
cd /usr/bin/
sudo unlink perl
sudo mv perl_ori perl
```

I can propose a fix in the doc for the first issue. As for the second (perl) issue, If someone has a simpler fix that would be great. Else I can document it as well in the doc so other people will have an easier time building the sdk.

I am able to generate complete portable tgz of Natron with centos7.
